frontends/concrete-python/concrete/fhe/mlir/context.py

```diff
@@ -238,29 +243,660 @@ def operation(
+                invert_accept = True
+
+        if invert_accept:
+            new_accept = set()
+            if Comparison.EQUAL in accept:
+                new_accept.add(Comparison.EQUAL)
+            if Comparison.LESS in accept:
+                new_accept.add(Comparison.GREATER)
+            if Comparison.GREATER in accept:
+                new_accept.add(Comparison.LESS)
+            accept = new_accept
+
+        intermediate_type = self.tensor(intermediate_scalar_type, resulting_type.shape)
+        lhs_minus_rhs = self.sub(intermediate_type, lhs, rhs)
+
+        accept_equal = int(Comparison.EQUAL in accept)
+        accept_greater = int(Comparison.GREATER in accept)
+        accept_less = int(Comparison.LESS in accept)
+
+        all_cells = 2**intermediate_bit_width
+
+        equal_cells = 1
+        greater_cells = (2 ** (intermediate_bit_width - 1)) - 1
+        less_cells = 2 ** (intermediate_bit_width - 1)
+
+        if not lhs_minus_rhs_dtype.is_signed:
+            lhs_minus_rhs = self.to_unsigned(lhs_minus_rhs)
+            greater_cells += less_cells
+            less_cells = 0
+
+        assert equal_cells + greater_cells + less_cells == all_cells
+
+        return self.tlu(
+            resulting_type,
+            lhs_minus_rhs,
+            (
+                [accept_equal] * equal_cells
+                + [accept_greater] * greater_cells
+                + [accept_less] * less_cells
+            ),
+        )
+
+    def comparison_with_chunks(
         self,
         resulting_type: ConversionType,
-        bit_width: int,
-        chunk_size: int,
         x: Conversion,
         y: Conversion,
+        accept: Set[Comparison],
+    ):
+        """
+        Compare encrypted values using chunks.
+
+        Idea:
+            split x and y into small chunks
+            compare the chunks using table lookups
+            reduce chunk comparisons to a final result
+        """
+
+        x_offset = 0
+        y_offset = 0
+
+        x_was_signed = x.is_signed
+        y_was_signed = y.is_signed
+
+        if x.is_signed or y.is_signed:
+            if x.is_signed:
+                signed_offset = 2 ** (x.original_bit_width - 1)
+                sanitizer = self.constant(self.i(x.bit_width + 1), signed_offset)
+                x = self.to_unsigned(self.add(x.type, x, sanitizer))
+                y_offset += signed_offset
+            if y.is_signed:
+                signed_offset = 2 ** (y.original_bit_width - 1)
+                sanitizer = self.constant(self.i(y.bit_width + 1), signed_offset)
+                y = self.to_unsigned(self.add(y.type, y, sanitizer))
+                x_offset += signed_offset
+
+        min_offset = min(x_offset, y_offset)
+
+        x_offset -= min_offset
+        y_offset -= min_offset
+
+        chunk_ranges = self.best_chunk_ranges(x, x_offset, y, y_offset)
+        assert 1 <= len(chunk_ranges) <= 3
+
+        if accept in ({Comparison.EQUAL}, {Comparison.LESS, Comparison.GREATER}):
```

Comment:
should be split into 
comparison_with_chunks_equals
comparison_with_chunks_generic